### PR TITLE
Add MMS CTC Fine-Tuning

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds language adapter fine-tuning for MMS. Still playing around with good hyper-parameters but script is functional.

Getting some very nice results now for:
```py
export CUDA_VISIBLE_DEVICES="0"
LEARNING_RATE="1e-3"

python run_speech_recognition_ctc.py \
        --dataset_name="common_voice" \
        --model_name_or_path="facebook/mms-1b-all" \
        --dataset_config_name="tr" \
        --output_dir="./wav2vec2-common_voice-tr-mms-demo" \
        --overwrite_output_dir \
        --num_train_epochs="15" \
        --per_device_train_batch_size="32" \
        --learning_rate="${LEARNING_RATE}" \
        --warmup_steps="400" \
        --evaluation_strategy="steps" \
        --text_column_name="sentence" \
        --length_column_name="input_length" \
        --save_steps="400" \
        --eval_steps="200" \
        --layerdrop="0.0" \
        --save_total_limit="3" \
        --adapter_attn_dim="16" \
        --adapter_language="tur" \
        --gradient_checkpointing \
        --chars_to_ignore , ? . ! - \; \: \" “ % ‘ ” � \
        --fp16 \
        --group_by_length \
        --do_train --do_eval
```

WER drops to 25% just after 200 steps. 
See: https://wandb.ai/patrickvonplaten/huggingface/runs/6f5cx5gg?workspace=user-patrickvonplaten

@sgugger @amyeroberts @sanchit-gandhi it'd be super nice to get a quick review here whether the code changes are generally fine with you. I'll only have to fill out the TODOs in the README with a nice example code and some description.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 1]
```diff
@@ -579,12 +632,24 @@ def remove_special_characters(batch):
         cache_dir=model_args.cache_dir,
         config=config,
         use_auth_token=data_args.use_auth_token,
+        ignore_mismatched_sizes=True,
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:635]: This is needed when instantiating from CTC checkpoints

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 2]
```diff
@@ -31,6 +31,7 @@
 import numpy as np
 import torch
 from datasets import DatasetDict, load_dataset
+from safetensors.torch import save_file as safe_save_file
```
Comment (patrickvonplaten) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:34]: It's a required dependency so should be fine

**Review by sgugger — COMMENTED** [review]
I think this should go in its own example instead of adding some more code to the (already complex) ctc example. It's preferable to have multiple examples focused on one thing than one big multi-purpose example.

**Comment by patrickvonplaten** [comment]
> I think this should go in its own example instead of adding some more code to the (already complex) ctc example. It's preferable to have multiple examples focused on one thing than one big multi-purpose example.

Ok for me

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! Changes all LGTM (filled in examples conditional ;))

+1 to @sgugger's suggestion of having a separate example

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 3]
```diff
@@ -132,6 +134,12 @@ class ModelArguments:
     ctc_loss_reduction: Optional[str] = field(
         default="mean", metadata={"help": "The way the ctc loss should be reduced. Should be one of 'mean' or 'sum'."}
     )
+    adapter_attn_dim: int = field(
+        default=None,
+        metadata={
+            "help": "If defined, adapter layers will be randomely initialized and the rest of the model will be frozen."
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:140]:
```suggestion
            "help": "If defined, adapter layers will be randomly initialized and the rest of the model will be frozen."
```

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 4]
```diff
@@ -261,6 +269,18 @@ class DataTrainingArguments:
             )
         },
     )
+    adapter_language: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": (
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:275]: Nice help message!

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 5]
```diff
@@ -132,6 +134,12 @@ class ModelArguments:
     ctc_loss_reduction: Optional[str] = field(
         default="mean", metadata={"help": "The way the ctc loss should be reduced. Should be one of 'mean' or 'sum'."}
     )
+    adapter_attn_dim: int = field(
+        default=None,
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:138]: Can this default to some sensible value or should we always force the user to pass it?

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py [hunk 6]
```diff
@@ -579,12 +632,24 @@ def remove_special_characters(batch):
...
     if model_args.freeze_feature_encoder:
         model.freeze_feature_encoder()
 
+    # if attn adapter is defined, freeze all non-adapter weights
+    if model.config.adapter_attn_dim is not None:
+        model.init_adapter_layers()
+        # first we freeze the whole base model
+        model.freeze_base_model()
+
+        # next we unfreeze all adapter layers
```
Comment (sanchit-gandhi) [inline examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:648]: Do we need to unfreeze the adapter weights? They don't get frozen in `model.freeze_base_model()`

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py (hunk 1 above)
Comment (sanchit-gandhi) [reply examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:635]: Good catch

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 7]
```diff
@@ -1194,6 +1194,19 @@ def _get_adapters(self):
 
         return adapter_weights
 
+    def init_adapter_layers(self):
```
Comment (sanchit-gandhi) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:1197]: Nice!

**Review by sanchit-gandhi — APPROVED** [review]
Looks good already, thanks for the updates @patrickvonplaten. Just some minor suggestions

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py (hunk 6 above)
Comment (patrickvonplaten) [reply examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:648]: They do get frozen in `model.freeze_base_model()` (adapter attention weights are part of it)

File: examples/pytorch/speech-recognition/run_speech_recognition_ctc.py (hunk 5 above)
Comment (patrickvonplaten) [reply examples/pytorch/speech-recognition/run_speech_recognition_ctc.py:138]: Yes good point, now that things will be moved to a new file, I'll set a good default

**Review by sgugger — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Added a test. Moved the code into a new example file. Added an extensive README. WER for a quick 10min run can be as low as 23% WER!

**Comment by patrickvonplaten** [comment]
Demo training run: https://huggingface.co/patrickvonplaten/wav2vec2-common_voice-tr-mms-demo

**Comment by dash8x** [comment]
In which release will this be available in?

**Comment by sanchit-gandhi** [comment]
You can find the examples scripts here: https://github.com/huggingface/transformers/tree/main/examples/pytorch/speech-recognition#connectionist-temporal-classification-with-adapters

They assume that you are running from the latest dev version:
https://github.com/huggingface/transformers/blob/f10452271802573fe6e19442631113c4c23a2c70/examples/pytorch/speech-recognition/run_speech_recognition_ctc_adapter.py#L55-L56

Which you can do by following the instructions for installing from source or editable install here: https://huggingface.co/docs/transformers/installation#install-from-source

Although for MMS ASR fine-tuning, you can safely run the script using the latest PyPi release version (4.31.0).

**Pull request merged**